src/roles.js

```diff
@@ -270,31 +276,156 @@ export class Roles extends TypedEmitter {
    * @returns {Promise<Role>}
    */
   async getRole(deviceId) {
-    const roleRecord = await this.#dataType
+    const membershipRecord = await this.#getMembershipRecord(deviceId)
+
+    switch (membershipRecord) {
+      case null:
+        return NO_ROLE
+      case CREATOR_MEMBERSHIP_RECORD:
+        return CREATOR_ROLE
+      default: {
+        const { roleId } = membershipRecord
+        if (
+          isRoleId(roleId) &&
+          (await this.#isRoleChainValid(membershipRecord))
+        ) {
+          return ROLES[roleId]
+        } else {
+          return BLOCKED_ROLE
+        }
+      }
+    }
+  }
+
+  /**
+   * @param {string} deviceId
+   * @returns {Promise<null | typeof CREATOR_MEMBERSHIP_RECORD | MembershipRecord>}
+   */
+  async #getMembershipRecord(deviceId) {
+    const result = await this.#dataType
       .getByDocId(deviceId)
       .catch(nullIfNotFound)
-    if (!roleRecord) {
-      // The project creator will have the creator role
-      const authCoreId = await this.#coreOwnership.getCoreId(deviceId, 'auth')
-      if (authCoreId === this.#projectCreatorAuthCoreId) {
-        return CREATOR_ROLE
-      } else {
-        // When no role assignment exists, e.g. a newly added device which has
-        // not yet synced role records.
-        return NO_ROLE
+    if (result) return result
+
+    // The project creator will have the creator role
+    const authCoreId = await this.#coreOwnership.getCoreId(deviceId, 'auth')
+    if (authCoreId === this.#projectCreatorAuthCoreId) {
+      return CREATOR_MEMBERSHIP_RECORD
+    }
+
+    // When no role assignment exists, e.g. a newly added device which has
+    // not yet synced membership records.
+    return null
+  }
+
+  /**
+   * @param {ReadonlyDeep<MembershipRecord>} membershipRecord
+   * @returns {Promise<boolean>}
+   */
+  async #isRoleChainValid(membershipRecord) {
+    if (membershipRecord.roleId === LEFT_ROLE_ID) return true
+
+    /** @type {null | ReadonlyDeep<MembershipRecord>} */
+    let currentMembershipRecord = membershipRecord
+
+    while (currentMembershipRecord) {
+      const parentMembershipRecord = await this.#getParentMembershipRecord(
+        currentMembershipRecord
+      )
+      switch (parentMembershipRecord) {
+        case null:
```

Comment:
Maybe be more explicit and return `false` here, since that is what will happen anyway.